stm32/libcanard/bxcan/src/bxcan.c

```diff
@@ -0,0 +1,763 @@
+        {
+            const bool tme[3U] = {(bxcan_base->TSR & BXCAN_TSR_TME0) != 0U,
+                                  (bxcan_base->TSR & BXCAN_TSR_TME1) != 0U,
+                                  (bxcan_base->TSR & BXCAN_TSR_TME2) != 0U};
+
+            for (uint8_t i = 0U; i < 3U; i++)
+            {
+                // TX deadline management.
+                // Abort the selected mailbox if it is busy and its deadline has expired. This frees-up stale
+                // slots and the timeout error is reported.
+                // However, if the slot was not busy, we only reset the deadline to zero, as this only indicates
+                // a previously-sent, idle slot that needs its old deadline zeroed. This is not an error.
+                const uint8_t tx_deadline_idx = i + (iface_index * 3);
+                if ((current_time > g_tx_deadline[tx_deadline_idx]) && (g_tx_deadline[tx_deadline_idx] != 0U))
+                {
+                    // If the slot is active, abort the transmission and report a time-out error.
+                    if (!tme[i])
+                    {
+                        // Abort the selected mailbox.
+                        if (i == 0U)
+                        {
+                            bxcan_base->TSR = BXCAN_TSR_ABRQ0;
+                        }
+                        if (i == 1U)
+                        {
+                            bxcan_base->TSR = BXCAN_TSR_ABRQ1;
+                        }
+                        if (i == 2U)
+                        {
+                            bxcan_base->TSR = BXCAN_TSR_ABRQ2;
+                        }
+
+                        *error_iface = true;  // TX timeout error occurred.
+                    }
+                    g_tx_deadline[tx_deadline_idx] = 0U;  // Clear the TX deadline for the selected mailbox.
+                }
+
+                if (tme[i])  // This TX mailbox is free, we can use it
+                {
+                    tx_mailbox = i;
+                }
+                else  // This TX mailbox is pending, check for priority inversion
+                {
+                    if (extended_can_id >= convertRegisterToFrameID(bxcan_base->TxMailbox[i].TIR))
+                    {
+                        // There's a mailbox whose priority is higher or equal the priority of the new frame.
+                        // Priority inversion would occur! Reject transmission. We do this by insuring that
+                        // the tx_mailbox value is set to the trip value of 0xFF and break the search loop.
+                        // The priority inversion logic then catches the condition & handles it.
+#if BXCAN_CONFIG_DEBUG_INNER_PRIORITY_INVERSION
+                        BXCAN_ASSERT(!"CAN PRIO INV");  // NOLINT
+#endif
+                        tx_mailbox = 0xFF;
+                        break;
+                    }
+                }
```

Comment:
I'm sorry but my previous advice was not entirely correct. I suggested that you move this section into the loop, but the correct action would be to move this fragment past the loop. The idea is that we check all TX entries first, and then, if none of them are free AND none of them have a higher priority, we detect the priority inversion: https://github.com/UAVCAN/libcanard/blob/8ee343c4edae0e0e4e1c040852aa3d8430f7bf76/drivers/stm32/canard_stm32.c#L323-L352

The logic is explained in the currently misplaced comment below ("If all TX mailboxes are busy...").

Reply:
I've re-worked it as I understand it now. 

The happy, little, compact code of v0 has really mutated. 855 lines of text and there assuredly still be dragons.